### PR TITLE
Support for SE mappers in MP.

### LINKED ISSUE #2090: SE Config mapping issue in MP
# Environment Details
* Helidon Version:  2.0.0
* Helidon  MP
----------

## Problem Description
* Helidon SE Config injected in Helidon MP into a CDI bean does not honor mappers on classpath, such as Object Mapping

### PR DESCRIPTION
**Pull request by tomas-langer**
Resolves #2090 

Signed-off-by: Tomas Langer <tomas.langer@oracle.com>

### CODE REVIEW

File: config/config-mp/src/main/java/io/helidon/config/mp/SeConfigValue.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import io.helidon.config.Config;
+import io.helidon.config.ConfigMappingException;
+import io.helidon.config.ConfigValue;
+import io.helidon.config.MissingValueException;
+
+class SeConfigValue<T> implements ConfigValue<T> {
```
Comment (ljnelson) [inline config/config-mp/src/main/java/io/helidon/config/mp/SeConfigValue.java:28]: Do you need `equals` and `hashCode` overrides for these `ConfigValue` implementations?  Or no?

File: config/config-mp/src/main/java/io/helidon/config/mp/SeConfig.java [hunk 2]
```diff
@@ -355,9 +386,23 @@ private MpConfigImpl impl() {
                     break;
                 }
             }
-            return Optional.of(result);
+            return new SeConfigValue<>(key, result);
         } else {
-            return Optional.empty();
+            // and further still we may have a list of objects
+            if (get("0").type() == Type.MISSING){
```
[the remark this answers is not in the document]
Comment (tomas-langer) [reply config/config-mp/src/main/java/io/helidon/config/mp/SeConfig.java:392]: Yes - if there is an array in configuration, it uses index. I am just checking that there is a subkey with the first index.
```
list.0=first
list.1=second
objectlist.0.name=first
objectlist.0.value=first value
objectlist.1.name=second
objectlist.1.value=second value
```

File: config/config-mp/src/main/java/io/helidon/config/mp/MpConfig.java [hunk 3]
```diff
@@ -40,12 +44,32 @@ private MpConfig() {
...
+            }
+        }
+
+        if (allSources.size() == 1) {
+            ConfigSource first = allSources.get(0);
+            if (first instanceof MpHelidonConfigSource) {
+                // we only have Helidon SE config as a source - let's just use it
+                return ((MpHelidonConfigSource) first).unwrap();
+            }
+        }
```
[the remark this answers is not in the document]
Comment (danielkec) [reply config/config-mp/src/main/java/io/helidon/config/mp/MpConfig.java:70]: Can be even shorter 
```java
    public static io.helidon.config.Config toHelidonConfig(Config mpConfig) {

        if (mpConfig instanceof io.helidon.config.Config) {
            return (io.helidon.config.Config) mpConfig;
        }

        return StreamSupport.stream(config.getConfigSources().spliterator(), false)
                .filter(MpHelidonConfigSource.class::isInstance)
                .findFirst()
                .map(MpHelidonConfigSource.class::cast)
                .map(MpHelidonConfigSource::unwrap)
                // we use Helidon SE config to handle object mapping (and possible other mappers on classpath)
                .orElseGet(() -> new SeConfig(io.helidon.config.Config.builder()
                        .sources(ConfigSources.empty())
                        .overrides(OverrideSources.empty())
                        .disableEnvironmentVariablesSource()
                        .disableSystemPropertiesSource()
                        .disableParserServices()
                        .disableFilterServices()
                        .disableCaching()
                        .disableValueResolving()
                        .changesExecutor(command -> {
                        })
                        .build(), mpConfig));
    }
```

File: config/config-mp/src/main/java/io/helidon/config/mp/SeConfig.java [hunk 4]
```diff
@@ -265,6 +275,27 @@ public String toString() {
...
+            return Stream.of(config);
+        } else {
+            return config.asNodeList()
+                    .map(list -> list.stream()
+                            .filter(predicate)
+                            .map(node -> traverseSubNodes(node, predicate))
+                            .reduce(Stream.of(config), Stream::concat))
+                    .orElseThrow(MissingValueException.createSupplier(key()));
+        }
+    }
```
[the remark this answers is not in the document]
Comment (danielkec) [reply config/config-mp/src/main/java/io/helidon/config/mp/SeConfig.java:297]: aa its not, missing type check needs to be even in the flatMap

File: config/config-mp/src/main/java/io/helidon/config/mp/MpConfig.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomas-langer) [reply config/config-mp/src/main/java/io/helidon/config/mp/MpConfig.java]: Fixed

File: config/config-mp/src/main/java/io/helidon/config/mp/SeConfigValue.java (hunk 1 above)
Comment (tomas-langer) [reply config/config-mp/src/main/java/io/helidon/config/mp/SeConfigValue.java:29]: Fixed

File: config/config-mp/src/main/java/io/helidon/config/mp/SeConfig.java [hunk 5]
```diff
@@ -355,9 +386,23 @@ private MpConfigImpl impl() {
...
+            // there are objects here, let's do that
+            List<T> result = new LinkedList<>();
+
+            for (int i = 0; i < 1000; i++) {
+                Config config = get(String.valueOf(i));
+                if (config.type() == Type.MISSING) {
+                    break;
+                }
+                result.add(config.as(typeArg).get());
+            }
```
Comment (danielkec) [inline config/config-mp/src/main/java/io/helidon/config/mp/SeConfig.java:407]: Magic number

File: config/config-mp/src/main/java/io/helidon/config/mp/SeConfig.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (danielkec) [reply config/config-mp/src/main/java/io/helidon/config/mp/SeConfig.java:297]: or does it?

File: config/config-mp/src/main/java/io/helidon/config/mp/SeConfig.java (hunk 5 above)
Comment (tomas-langer) [reply config/config-mp/src/main/java/io/helidon/config/mp/SeConfig.java:407]: very magical, but documented

Comment (tomas-langer) [reply config/config-mp/src/main/java/io/helidon/config/mp/SeConfig.java]: Fixed anyway

File: config/config-mp/src/main/java/io/helidon/config/mp/SeConfig.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (tomas-langer) [reply config/config-mp/src/main/java/io/helidon/config/mp/SeConfig.java:297]: I don't feel like changing this right now. If you feel this is a problem, let's do a follow up issue.

**Pull request merged**